spring-ws-core/src/main/java/org/springframework/ws/soap/saaj/SaajSoapMessageFactory.java

```diff
@@ -233,7 +235,19 @@ private MimeHeaders parseMimeHeaders(InputStream inputStream) throws IOException
 		MimeHeaders mimeHeaders = new MimeHeaders();
 		if (inputStream instanceof TransportInputStream) {
 			TransportInputStream transportInputStream = (TransportInputStream) inputStream;
-			for (Iterator<String> headerNames = transportInputStream.getHeaderNames(); headerNames.hasNext();) {
+			
+			//Remove duplicate keys
+			Iterator<String> headerNamesWithDuplicate = transportInputStream.getHeaderNames();
```

Comment:
I see. In that case, I'd rather see that fixed in HttpUrlConneciton.getResponseHeaderNames(), so that it filters out duplicate header names there, than here. The easiest way would be to use a Set, rather than a List, at line 105.